### ISSUE TITLE
RFC: Caching the different versions of the binary in wrapper tools

### ISSUE THREAD

**Issue by mstruebing** [editorconfig-checker/editorconfig-checker#97]
In every wrapper tool the release is currently downloaded to the directory where the wrapper tool is located and extracted to the `bin` directory right there.

We should consider caching the already downloaded binary, especially since maybe you have many projects which depend on it what would be a massive waste of space.

My suggestion would be (inspired and adopted from @vbrandl): 

1. Download the release in a writable temporary directory
2. Extract the binary in the temporary directory
3. Verify the correctness of the binary via some hashing function (maybe sha512?), maybe the right hash should be hard-coded into the wrapper itself?
4. Move the binary either to `$XDG_DATA_HOME/[ec|editorconfig-checker]/x.x.x/ec` `$XDG_CACHE_HOME/[ec|editorconfig-checker]/x.x.x/ec`- I'm not really sure what would be more sufficient.
5. Check the correctness of the binary on every execution of the wrapper, if hash doesn't match start with point 1.

Is there something I'm missing? Are there any other good ideas?
Feel free to tell me :)

**Comment by vbrandl** [comment]
Few notes/comments:

To 3: I wouldn't hard code the hash but put a `CHECKSUMS`/`SHA256_*` file, that contains the hashes for all released binaries, in the GitHub release for anyone to check. For even more integrity, the checksums file could be signed and the public key gets hard coded in the wrapper. That way you don't need to update the wrapper every time the checker is updated

To 4: Use `XDG_CACHE_HOME`. `XDG_DATA_HOME` is intended for user data and must not be deleted. `XDG_CACHE_HOME` can be deleted without any changes in behavior which fits for our use case. Also the binary itself should be stored as `ec-<OS>-<arch>` since in theory `XDG_CACHE_HOME` might be a network share and different OS use it as cache directory. If the OS and architecture is encoded in the file name, there can be no conflicts. The rest of the structure (`ec` vs `editorconfig-checker`) must be discussed and standardized.

To 5: With the changes proposed in 3 in mind, we should also download the checksum file so we don't need to make a HTTP request each time to verify the signature.

**Comment by mstruebing** [comment]
I agree with your points except for one part of point 3: 

The wrapper should be updated when the binary is updated.
If I'm using i.e v2.3 in a javascript project and run the checker and it updates automatically to v3.0 (which is some major change in behaviour) my project could be broken until I fix some things. I would rather want my $package_manager from my specific programming language to handle updates with version ranges or fully manually.

**Comment by vbrandl** [comment]
I agree with the breaking changes but I think it's too much organizational overhead to create a new release for _every_ wrapper, once there is an update for the checker.

How about pinning the checker version in `.ecrc`? This allows us to give feedback, if an update is available (e.g. by printing a warning to stderr) and update the checker without having to update the wrapper.

**Comment by mstruebing** [comment]
Good idea, I think this is the way to go, as releasing every wrapper was a bit of work to do everytime. Cool to have you and your ideas on board :)

**Comment by vbrandl** [comment]
Ok to formalize the RFC a bit:

We need new fields in .ecrc:

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| `version` | semver string | `latest` | Checker version to be used. Used to pin the checker version. Should support partial versions, e.g. `2` for the latest `2.*.*` version. If this is not set or set to `latest`, always use the latest available version  |
| `failOnUpdate` | bool | `false` | If an outdated checker version is a hard error (e.g. exit code 1) or just a warning (e.g. printing to `stderr`) |
| `checkInterval` | int | `1440` (1 day) | Number of minutes between update checks (so we don't need to check on every execution) |

Location for the checker binary: `$XDG_CACHE_HOME/editorconfig-checker/<semver>/ec-<os>-<arch>`

Metadata files:

| File | Function |
| --- | --- |
| `$XDG_CACHE_HOME/editorconfig-checker/latest` | Contains the latest version of the checker. Every time, the version is checked, the `last modified` date is set to `NOW`. In combination with `checkInterval` we can now determine if we need to check again |

*NOTE:* I don't know if there are file systems without a `last modified` field. If so, we need another mechanism for this.

**BUT**: If we are downloading dynamically we MUST be able to verify the integrity of the checker. Before this gets implemented, releases of the checker must contain a signed file with checksums. Otherwise, if an attacker gained access to the repository, it just needs to create a new release and it will be downloaded and executed.

**Comment by mstruebing** [comment]
Could you explain how we could create a signed file with checksums?
I'm assuming simply creating a sha512sum or something wouldn't be sufficient.

**Comment by vbrandl** [comment]
I was thinking of something like this:

* https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/SHA512SUMS
* https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/SHA512SUMS.sign

For the key, I've 2 ideas:

* single signing key: If another maintainer should be able to sign releases, you have to give out a copy of the signing key (doesn't seem like a good idea to me)
* master key: you generate a master key for the editorconfig-checker project as a whole and use this key to sign the maintainer's key
